Config validation error

whilefoo opened the pull request:
Resolves #680 

Example:
https://github.com/ubiquity-whilefoo/ubiquibot/commit/b74601f5b46603abe0e4b6d6799aaa10ea6404c7#comments
https://github.com/ubiquity-whilefoo/ubiquibot/commit/c8e02c01236b089d702c3b5511af86c5d4f781a3#comments

It was kinda hard to log config errors because if the config was invalid, I also didn't have Supabase keys which the logger needs to write logs to database, so I had to write additional fallbacks

pavlovcik commented:
First one is great. Second one is vague is there anything you can do to improve that type of message?

whilefoo commented:
I improved the message:
https://github.com/ubiquity-whilefoo/ubiquibot/commit/386a77299e1df422473177a7e144a6075ec330f6#comments

0xcodercrane commented on src/adapters/index.ts:
```diff
@@ -6,7 +6,7 @@ export * from "./telegram";
   return {
-    supabase: supabase(config.supabase.url, config.supabase.key),
-    telegram: new Telegraf(config.telegram.token).telegram,
+    supabase: supabase(config?.supabase?.url ?? process.env.SUPABASE_URL, config?.supabase?.key ?? process.env.SUPABASE_KEY),
+    telegram: new Telegraf(config?.telegram?.token ?? process.env.TELEGRAM_BOT_TOKEN).telegram,
```
It's not a good idea to put the env variables here. `env` variables should be read in [config.ts](https://github.com/ubiquity/ubiquibot/blob/development/src/bindings/config.ts)

0xcodercrane replied:
bumping @whilefoo here. lets follow up the general process to resolve the conversation. you should at least leave a comment.

0xcodercrane replied:
Even if config separation resolves the problem finally, we should fix it here as well.

whilefoo replied:
How can we fix it here? Config is not available if it's invalid

0xcodercrane replied:
yeah I mean if the config is not available here(maybe null), we don't need to put env variables for adapters because the bot instance would not work. If we have to forcely make the supabase working for logging, we can allow it only for supabase adapter but not for telegram adapter. Ideally we should remove it from supabase adapter as well in the future.

0xcodercrane replied:
ok lets make it simple. suggested code changes

0xcodercrane commented on src/bindings/event.ts:
```diff
@@ -43,11 +48,21 @@ export const bindEvents = async (context: Context): Promise<void> => {
     return;
   }
 
+  if (botConfigError) {
+    logger.error(botConfigError.toString());
+    return;
+  }
```
I guess these lines need to be moved to here.

line 38 - line 42

 ```
 try {
    botConfig = await loadConfig(context);
  } catch (err) {
    botConfigError = err;
  }
```

whilefoo replied:
ideally yes, but logger needs to be initialized for this to work, but logger gets initialized after loading the config because it needs config for supabase and other variables. but if the loading of the config fails then we don't have those variables - this is why I had to make additional fallbacks like [here](https://github.com/ubiquity/ubiquibot/pull/684#discussion_r1309506414)

0xcodercrane replied:
oh right.

0xcodercrane commented on src/handlers/push/index.ts:
```diff
@@ -45,3 +49,63 @@ export const runOnPush = async () => {
+            return error.params.additionalProperty;
+          }
+        });
+        await context.octokit.rest.repos.createCommitComment({
+          owner: payload.repository.owner.login,
+          repo: payload.repository.name,
+          commit_sha: commitSha,
+          body: `@${payload.sender.login} Config validation failed! Error: ${ajv.errorsText()}. ${
+            additionalProperties && additionalProperties.length > 0 ? `Unnecessary properties: ${additionalProperties.join(", ")}` : ""
+          }`,
+          path: BASE_RATE_FILE,
+        });
```
need to create a helper function in src/helper.ts for this and use the helper function here.

0xcodercrane commented on src/utils/private.ts:
```diff
@@ -47,49 +49,77 @@ export const getConfigSuperset = async (context: Context, type: "org" | "repo",
+    comment: Type.Object(
+      {
+        elements: Type.Record(Type.String(), Type.Number()),
+        totals: Type.Object(
+          {
+            word: Type.Number(),
+          },
+          { additionalProperties: false }
+        ),
+      },
+      { additionalProperties: false }
+    ),
+  },
+  { additionalProperties: false }
+);
+
+export type Incentives = Static<typeof IncentivesSchema>;
+
+export const WideConfigSchema = Type.Object(
+  {
+    "evm-network-id": Type.Optional(Type.Number()),
+    "price-multiplier": Type.Optional(Type.Number()),
+    "issue-creator-multiplier": Type.Number(),
+    "time-labels": Type.Optional(Type.Array(WideLabelSchema)),
+    "priority-labels": Type.Optional(Type.Array(WideLabelSchema)),
+    "payment-permit-max-price": Type.Optional(Type.Number()),
+    "command-settings": Type.Optional(Type.Array(CommandObjSchema)),
+    "promotion-comment": Type.Optional(Type.String()),
+    "disable-analytics": Type.Optional(Type.Boolean()),
+    "comment-incentives": Type.Optional(Type.Boolean()),
+    "assistive-pricing": Type.Optional(Type.Boolean()),
+    "max-concurrent-assigns": Type.Optional(Type.Number()),
+    incentives: Type.Optional(IncentivesSchema),
+    "default-labels": Type.Optional(Type.Array(Type.String())),
+    "register-wallet-with-verification": Type.Optional(Type.Boolean()),
+  },
+  {
+    additionalProperties: false,
+  }
+);
```
Looks good to replace `interfaces` with `schemas` and i guess it would be better for you to move the schemas to `src/types`.

0xcodercrane commented on src/utils/private.ts:
```diff
@@ -160,7 +190,13 @@ export const getWideConfig = async (context: Context) => {
   const repoConfig = await getConfigSuperset(context, "repo", CONFIG_PATH);
 
   const parsedOrg: WideOrgConfig | undefined = parseYAML(orgConfig);
+  if (parsedOrg !== undefined && !ajv.validate(WideOrgConfigSchema, parsedOrg)) {
```
```suggestion
  if (parsedOrg && !ajv.validate(WideOrgConfigSchema, parsedOrg)) {
```
same thing right?

0xcodercrane commented on src/utils/private.ts:
```diff
@@ -160,7 +190,13 @@ export const getWideConfig = async (context: Context) => {
+    throw new Error(`Invalid org config: ${ajv.errorsText()}`);
+  }
   const parsedRepo: WideRepoConfig | undefined = parseYAML(repoConfig);
+  if (parsedRepo !== undefined && !ajv.validate(WideConfigSchema, parsedRepo)) {
```
```suggestion
  if (parsedRepo && !ajv.validate(WideConfigSchema, parsedRepo)) {
```

web4er commented:
It would be nice to get the name additional properties in console logs and supabase logs.

```
biquiBot Error: Invalid repo config: data must NOT have additional properties undefined error ubiquibot web4erOrg undefined undefined
```

The bot didn't add comment here on wrong config. 
https://github.com/web4erOrg/ubiquibot/commit/e556157056ef7ab8ae9e7003b02d164277b4c31c


I have only `price-multiplier` and `issue-creator-multiplier` in my org config. If I add only `price-multiplier` in repo config, this shows up in console log and supabase log, but nothing in commit comments. 
```
UbiquiBot Error: Invalid repo config: data must have required property 'issue-creator-multiplier' undefined error ubiquibot web4erOrg undefined undefined
```

whilefoo commented:
> It would be nice to get the name additional properties in console logs and supabase logs.

Added

> The bot didn't add comment here on wrong config.
> [web4erOrg@e556157](https://github.com/web4erOrg/ubiquibot/commit/e556157056ef7ab8ae9e7003b02d164277b4c31c)

This happened because the config on development was invalid, so you pushed the change, the bot tried to load the current config but it's invalid so it stops. I modified the code to run validateConfigChange even if config load fails.

0xcodercrane commented on src/adapters/index.ts:
```diff
@@ -6,7 +6,7 @@ export * from "./telegram";
-    supabase: supabase(config.supabase.url, config.supabase.key),
-    telegram: new Telegraf(config.telegram.token).telegram,
+    supabase: supabase(config?.supabase?.url ?? process.env.SUPABASE_URL, config?.supabase?.key ?? process.env.SUPABASE_KEY),
+    telegram: new Telegraf(config?.telegram?.token ?? process.env.TELEGRAM_BOT_TOKEN).telegram,
```
```suggestion
    telegram: new Telegraf(config?.telegram?.token ?? "").telegram,
```

0xcodercrane commented on src/adapters/index.ts:
```diff
@@ -6,7 +6,7 @@ export * from "./telegram";
-    supabase: supabase(config.supabase.url, config.supabase.key),
-    telegram: new Telegraf(config.telegram.token).telegram,
+    supabase: supabase(config?.supabase?.url ?? process.env.SUPABASE_URL, config?.supabase?.key ?? process.env.SUPABASE_KEY),
+    telegram: new Telegraf(config?.telegram?.token ?? process.env.TELEGRAM_BOT_TOKEN).telegram,
```
```suggestion
    telegram: new Telegraf(config?.telegram?.token ?? "").telegram,
```

0xcodercrane replied:
will do it in `development` branch

0xcodercrane commented:
It looks like you should merge the `development` branch into your local `config-validation-error` branch @whilefoo
